python/dgl/distributed/dist_dataloader.py

```diff
@@ -0,0 +1,159 @@
+        """
+        assert num_workers > 0, "DistDataloader only supports num_workers>0 for now. if you \
+            want to use single process dataloader, please use PyTorch dataloader for now"
+        if queue_size is None:
+            queue_size = num_workers * 4
+        self.queue_size = queue_size
+        self.batch_size = batch_size
+        self.queue_size = queue_size
+        self.collate_fn = collate_fn
+        self.current_pos = 0
+        self.num_workers = num_workers
+        self.m = mp.Manager()
+        self.queue = self.m.Queue(maxsize=queue_size)
+        self.sig_queue = self.m.Queue(maxsize=num_workers)
+        self.drop_last = drop_last
+        self.send_idxs = 0
+        self.recv_idxs = 0
+        self.started = False
+        self.shuffle = shuffle
+
+        self.pool, num_sampler_workers = get_sampler_pool()
+        if self.pool is None:
+            ctx = mp.get_context("spawn")
+            self.pool = ctx.Pool(num_workers)
+        else:
+            assert num_sampler_workers == num_workers, "Num workers should be the same"
+        results = []
+        for _ in range(num_workers):
+            results.append(self.pool.apply_async(
+                init_fn, args=(collate_fn, self.queue, self.sig_queue)))
+            time.sleep(0.1)
+        for res in results:
+            res.get()
+
+        self.dataset = F.tensor(dataset)
+        self.expected_idxs = len(dataset) // self.batch_size
+        if not self.drop_last and len(dataset) % self.batch_size != 0:
+            self.expected_idxs += 1
+
+    def __next__(self):
+        if not self.started:
+            for _ in range(self.queue_size):
+                self._request_next_batch()
+        self._request_next_batch()
+        if self.recv_idxs < self.expected_idxs:
+            result = self.queue.get(timeout=9999)
+            self.recv_idxs += 1
+            return result
+        else:
+            self.recv_idxs = 0
+            self.current_pos = 0
+            raise StopIteration
+
+    def __iter__(self):
+        if self.shuffle:
+            self.dataset = F.rand_shuffle(self.dataset)
+        return self
+
+    def _request_next_batch(self):
+        next_data = self._next_data()
+        if next_data is None:
+            return None
+        else:
+            async_result = self.pool.apply_async(
+                call_collate_fn, args=(next_data, ))
+            self.send_idxs += 1
+            return async_result
+
+    def _next_data(self):
+        if self.current_pos == len(self.dataset):
+            return None
+
+        end_pos = 0
+        if self.current_pos + self.batch_size > len(self.dataset):
+            if self.drop_last:
+                return None
+            else:
+                end_pos = len(self.dataset)
+        else:
+            end_pos = self.current_pos + self.batch_size
+        ret = self.dataset[self.current_pos:end_pos]
+        self.current_pos = end_pos
+        return ret
+
+    def close(self):
```

Comment:
do users have to call it explicitly? should we move the code to `__del__`?

Reply:
User doesn't need to call this. However, in the test the DistDataloader object is in the main process and I also `join` the server in the main process, therefore cannot avoid explicitly call exit_client and close in the test.

Reply:
if this is just for testing, i think it's better to remove it. DistTensor is exposed to users. we don't want users to call the API.

Reply:
https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L492
As mentioned in the docs above, __del__ is not guaranteed to be called.

Reply:
I tested the exit_client function is called by atexit, in both main process and sampler process